processor/src/host/debug.rs

```diff
@@ -39,14 +82,173 @@ fn print_vm_stack(clk: u32, stack: Vec<Felt>, n: usize) {
+#[cfg(feature = "std")]
+fn print_mem_all(clk: u32, ctx: u32, mem: Vec<(u64, Word)>) {
+    println!("Memory state before step {clk} for the context {ctx}:");
+
+    // calculate the proper padding for the memory addresses
+    let padding = mem.iter().fold(0, |max, value| value.0.to_string().len().max(max));
+
+    // print the main part of the memory (wihtout the last value)
+    for (addr, value) in mem.iter().take(mem.len() - 1) {
+        println!(
+            "├── {addr:>width$}: [{}, {}, {}, {}]",
+            value[0],
+            value[1],
+            value[2],
+            value[3],
+            width = padding
+        );
+    }
+
+    // print the last memory value
+    if let Some((addr, value)) = mem.last() {
+        println!(
+            "└── {addr:>width$}: [{}, {}, {}, {}]\n",
+            value[0],
+            value[1],
+            value[2],
+            value[3],
+            width = padding
+        );
+    } else {
+        println!("└── EMPTY\n");
+    }
+}
+
+/// Prints the memory value stored at the requested address `addr`.
+#[cfg(feature = "std")]
+fn print_mem(clk: u32, ctx: u32, addr: u32, value: Option<Word>, local: bool) {
+    if local {
+        println!("Local state before step {} for the context {} at index {}:", clk, ctx, addr);
+    } else {
+        println!("Memory state before step {} for the context {} at address {}:", clk, ctx, addr);
+    }
+
+    if let Some(value) = value {
+        println!("└── [{}, {}, {}, {}]\n", value[0], value[1], value[2], value[3],);
+    } else {
+        println!("└── EMPTY\n");
+    }
+}
+
+/// Prints locals in provided interval.
+///
+/// If `print_all` is true, the interval should contain all indexes of locals available for the
+/// current procedure.
+#[cfg(feature = "std")]
+fn print_local_interval<S: ProcessState>(
+    process: &S,
+    clk: u32,
+    ctx: u32,
+    interval: (u32, u32),
+    print_all: bool,
+) {
+    let mut local_mem_interval = Vec::new();
+    let local_memory_offset = 2u32.pow(30) + 1;
+    for index in interval.0..interval.1 + 1 {
+        local_mem_interval.push((index, process.get_mem_value(ctx, index + local_memory_offset)))
+    }
+    if print_all {
+        print_mem_interval(
+            local_mem_interval,
+            format!("Local state before step {clk} for the context {ctx}:"),
+        );
+    } else {
+        print_mem_interval(
+            local_mem_interval,
+            format!(
+                "Local state before step {clk} for the context {ctx} in the interval [{}, {}]:",
+                interval.0, interval.1
+            ),
+        );
+    }
+}
+
+/// Prints memory values in the provided addresses interval.
+///
+/// `header` contains a title message corresponding to the DebugOption it is used in
+/// ([DebugOptions::MemInterval] or [DebugOptions::LocalInterval]).
+#[cfg(feature = "std")]
+fn print_mem_interval(mem: Vec<(u32, Option<Word>)>, header: String) {
```

Comment:
To simplify things, I'd probably not have separate functions for handling individual address and intervals. We can probably treat a single address as a special case of intervals.